dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,179 @@
+import requests
+import urllib3
+
+from dlab_core.domain.exceptions import DLabException
+from dlab_core.domain.helper import validate_property_type
+
+from jose import jwt
+
+
+URL_INTROSPECT = '{realm_address}/protocol/openid-connect/token/introspect'
+
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+
+class KeyCloakException(DLabException):
+    pass
+
+
+class KeyCloakConnectionException(KeyCloakException):
+    pass
+
+
+class StringProperty(object):
+    """A descriptor that forbids non-string values """
+
+    def __init__(self, value=None):
+        self.value = value
+
+    def __get__(self, instance, owner):
+        return self.value
+
+    @validate_property_type(str, arg_index=2)
+    def __set__(self, instance, value):
+        self.value = value
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    client_id = StringProperty()
+    client_secret = StringProperty()
+    realm_name = StringProperty()
+    keycloak_host = StringProperty()
+
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.realm_name = realm_name
+        self.keycloak_host = keycloak_host
+        self.realm_address = '{}/realms/{}'.format(
+            self.keycloak_host.rstrip('/'), self.realm_name
+        )
+        self._pub_key = None
+
+    @property
+    def public_key(self):
+        if not self._pub_key:
+            self._pub_key = self.retrieve_key()
+
+        return self._pub_key
+
+    def make_request_to_server(self, method, endpoint, verify=False, **kwargs):
+        """Performs request to endpoint
+
+        :type method: str
+        :param method: method for the HTTP request
+        :type endpoint: str
+        :param endpoint: endpoint to make request to
+        :type verify: bool
+        :param verify: whether to verify server's TLS certificate
+
+        :rtype: requests.Response
+        :return: HTTP response object
+        """
+        try:
+            response = requests.request(
+                method, endpoint, verify=verify, **kwargs
+            )
+            response.raise_for_status()
+            return response
+        except requests.exceptions.RequestException as e:
+            raise KeyCloakConnectionException(str(e))
+
+    def retrieve_key(self):
+        """Retrieves public key from KeyCloak realm
+
+        :rtype: str
+        :return: public key for token decoding
+        """
+        response = self.make_request_to_server(
```

Comment:
https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas

Reply:
Added commas